### PR TITLE
Move sets member propagation to SolverState

### PR DESCRIPTION
**Pull request by ajreynol**
This eliminates the parent relationship from solver state to theory sets.

### CODE REVIEW

File: src/theory/sets/solver_state.cpp [hunk 1]
```diff
@@ -469,6 +467,126 @@ const vector<Node> SolverState::getSetsEqClasses(const TypeNode& t) const
...
+void SolverState::addMember(TNode r, TNode atom)
+{
+  NodeIntMap::iterator mem_i = d_members.find(r);
+  size_t n_members = 0;
+  if (mem_i != d_members.end())
+  {
+    n_members = (*mem_i).second;
+  }
+  d_members[r] = n_members + 1;
+  if (n_members < d_members_data[r].size())
```
Comment (mudathirmahgoub) [inline src/theory/sets/solver_state.cpp:501]: When do we need this if statement? Isn't `n_numbers >= d_members_data[r].size` and invariant?

File: src/theory/sets/solver_state.cpp [hunk 2]
```diff
@@ -469,6 +467,126 @@ const vector<Node> SolverState::getSetsEqClasses(const TypeNode& t) const
...
+        {
+          // conflict
+          Assert(facts.empty());
+          Trace("sets-prop")
+              << "Propagate eq-mem conflict : " << exp << std::endl;
+          facts.push_back(exp);
+          return false;
+        }
+      }
+      if (n_members < d_members_data[t1].size())
```
Comment (mudathirmahgoub) [inline src/theory/sets/solver_state.cpp:575]: same here

File: src/theory/sets/theory_sets_private.cpp [hunk 3]
```diff
@@ -126,71 +125,27 @@ void TheorySetsPrivate::eqNotifyMerge(TNode t1, TNode t2)
     }
     // merge membership list
     Trace("sets-prop-debug") << "Copying membership list..." << std::endl;
-    NodeIntMap::iterator mem_i2 = d_members.find(t2);
-    if (mem_i2 != d_members.end())
+    // if t1 has a singleton and s2 does not, we may have new inferences to
```
Comment (mudathirmahgoub) [inline src/theory/sets/theory_sets_private.cpp:128]: typo `if s1`

File: src/theory/sets/solver_state.cpp [hunk 4]
```diff
@@ -469,6 +467,126 @@ const vector<Node> SolverState::getSetsEqClasses(const TypeNode& t) const
...
+      }
+    }
+    if (add)
+    {
+      // if there is a concrete set in t1, propagate new facts or conflicts
+      if (!cset.isNull())
+      {
+        NodeManager* nm = NodeManager::currentNM();
+        Assert(m2[1].getType().isComparableTo(cset.getType()));
+        Assert(areEqual(m2[1], cset));
```
Comment (mudathirmahgoub) [inline src/theory/sets/solver_state.cpp:552]: It seems to me this assertion is stronger than the assertion in the previous line which can be removed.

**Review by mudathirmahgoub — APPROVED** [review]
Thanks for the removing setParent function. Please address the comments before merge.

File: src/theory/sets/solver_state.cpp (hunk 1 above)
Comment (ajreynol) [reply src/theory/sets/solver_state.cpp:501]: No, n_members is context dependent and d_members_data is not, this means that n_members can backtrack to a smaller value.

**Pull request merged**